trieste/acquisition/rule.py

```diff
@@ -1097,6 +1104,225 @@ def state_func(
+        intersection of the trust region and ``search_space``.
+
+        :param search_space: The local acquisition search space for *this step*.
+        :param models: The model for each tag.
+        :param datasets: The known observer query points and observations. Uses the data for key
+            `OBJECTIVE` to calculate the new trust region.
+        :return: A function that constructs the next acquisition state and the recommended query
+            points from the previous acquisition state.
+        :raise KeyError: If ``datasets`` does not contain the key `OBJECTIVE`.
+        """
+
+        if models.keys() != {OBJECTIVE}:
+            raise ValueError(
+                f"dict of models must contain the single key {OBJECTIVE}, got keys {models.keys()}"
+            )
+
+        if datasets is None or datasets.keys() != {OBJECTIVE}:
+            raise ValueError(
+                f"""datasets must be provided and contain the single key {OBJECTIVE}"""
+            )
+
+        dataset = datasets[OBJECTIVE]
+        model = models[OBJECTIVE]
+
+        global_lower = search_space.lower
+        global_upper = search_space.upper
+
+        y_min = tf.reduce_min(dataset.observations, axis=0)
+
+        def state_func(
+            state: TURBO.State | None,
+        ) -> tuple[TURBO.State | None, TensorType]:
+            if state is None:  # initialise first TR
+                L, failure_counter, success_counter = self._L_init, 0, 0
+            else:  # update TR
+                step_is_success = y_min < state.y_min - 1e-10  # maybe make this stronger?
+                failure_counter = (
+                    0 if step_is_success else state.failure_counter + 1
+                )  # update or reset counter
+                success_counter = (
+                    state.success_counter + 1 if step_is_success else 0
+                )  # update or reset counter
+                L = state.L
+                if success_counter == self._success_tolerance:
+                    L *= 2.0  # make region bigger
+                    success_counter = 0
+                elif failure_counter == self._failure_tolerance:
+                    L *= 0.5  # make region smaller
+                    failure_counter = 0
+
+                L = tf.minimum(L, self._L_max)
+                if L < self._L_min:  # if gets too small then start again
+                    L, failure_counter, success_counter = self._L_init, 0, 0
+
+            # build region with volume according to length L but stretched according to lengthscales
+            xmin = dataset.query_points[tf.argmin(dataset.observations)[0], :]  # centre of region
+            lengthscales = (
+                model.get_kernel().lengthscales
+            )  # stretch region according to model lengthscales
+            tr_width = (
+                lengthscales * L / tf.reduce_prod(lengthscales) ** (1.0 / global_lower.shape[-1])
```

Comment:
Not necessary but a bit easier to read?

```suggestion
                lengthscales * L / (tf.reduce_prod(lengthscales) ** (1.0 / global_lower.shape[-1]))
```